File: code/packages/python/graph-native/tests/test_graph.py
from __future__ import annotations

import pytest

from graph_native import (
    EdgeNotFoundError,
    Graph,
    GraphRepr,
    NodeNotFoundError,
    bfs,
    connected_components,
    dfs,
    has_cycle,
    is_connected,
    minimum_spanning_tree,
    shortest_path,
)


def make_graph(repr: GraphRepr) -> Graph:
    graph = Graph(repr=repr)
    graph.add_edge("London", "Paris", 300.0)
    graph.add_edge("London", "Amsterdam", 520.0)
    graph.add_edge("Paris", "Berlin", 878.0)
    graph.add_edge("Amsterdam", "Berlin", 655.0)
    graph.add_edge("Amsterdam", "Brussels", 180.0)
    return graph


@pytest.fixture(params=[GraphRepr.ADJACENCY_LIST, GraphRepr.ADJACENCY_MATRIX])
def repr_kind(request: pytest.FixtureRequest) -> GraphRepr:
    return request.param


def test_node_and_edge_operations(repr_kind: GraphRepr) -> None:
    graph = Graph(repr=repr_kind)
    graph.add_node("A")
    graph.add_edge("A", "B", 2.5)

    assert graph.has_node("A")
    assert graph.has_node("B")
    assert graph.has_edge("A", "B")
    assert graph.has_edge("B", "A")
    assert graph.edge_weight("A", "B") == pytest.approx(2.5)
    assert graph.nodes() == frozenset({"A", "B"})
    assert graph.edges() == frozenset({("A", "B", 2.5)})
    assert graph.degree("A") == 1


def test_remove_edge_and_node_errors(repr_kind: GraphRepr) -> None:
    graph = Graph(repr=repr_kind)
    graph.add_edge("A", "B")
    graph.remove_edge("A", "B")
    graph.remove_node("B")

    with pytest.raises(EdgeNotFoundError):
        graph.remove_edge("A", "B")
    with pytest.raises(NodeNotFoundError):
        graph.neighbors("B")


def test_neighbors_weighted_and_repr(repr_kind: GraphRepr) -> None:
    graph = make_graph(repr_kind)
    assert graph.neighbors("Amsterdam") == frozenset({"London", "Berlin", "Brussels"})
    assert graph.neighbors_weighted("Amsterdam") == {
        "Berlin": 655.0,
        "Brussels": 180.0,
        "London": 520.0,
    }
    assert "Graph(" in repr(graph)


def test_traversals_and_connectivity(repr_kind: GraphRepr) -> None:
    graph = make_graph(repr_kind)
    assert bfs(graph, "London") == [
        "London",
        "Amsterdam",
        "Paris",
        "Berlin",
        "Brussels",
    ]
    assert dfs(graph, "London")[0] == "London"
    assert is_connected(graph)


def test_components_cycle_shortest_path_and_mst(repr_kind: GraphRepr) -> None:
    graph = make_graph(repr_kind)
    other = Graph(repr=repr_kind)
    other.add_edge("A", "B")
    other.add_edge("B", "C")
    other.add_edge("C", "A")
    other.add_edge("D", "E")

    components = connected_components(other)
    assert frozenset({"A", "B", "C"}) in components
    assert frozenset({"D", "E"}) in components
    assert has_cycle(other)

    assert shortest_path(graph, "London", "Berlin") == [
        "London",
        "Amsterdam",
        "Berlin",
    ]

    mst = minimum_spanning_tree(graph)
    assert len(mst) == len(graph) - 1
    assert sum(weight for _, _, weight in mst) == pytest.approx(1655.0)


def test_disconnected_mst_raises(repr_kind: GraphRepr) -> None:
    graph = Graph(repr=repr_kind)
    graph.add_edge("A", "B")
    graph.add_node("C")

    with pytest.raises(ValueError, match="not connected"):
        minimum_spanning_tree(graph)


def test_instance_algorithm_methods_and_dunders() -> None:
    graph = Graph()
    graph.add_edge("A", "B")
    graph.add_edge("B", "C")

    assert len(graph) == 3
    assert "A" in graph
    assert 7 not in graph
    assert graph.bfs("A") == ["A", "B", "C"]
    assert graph.dfs("A") == ["A", "B", "C"]
    assert graph.is_connected()
    assert graph.connected_components() == [frozenset({"A", "B", "C"})]
    assert not graph.has_cycle()
    assert graph.shortest_path("A", "C") == ["A", "B", "C"]
    assert graph.minimum_spanning_tree() == frozenset(
        {("A", "B", 1.0), ("B", "C", 1.0)}
    )

File: code/packages/lua/graph/tests/test_graph.lua
local script = arg and arg[0] or "tests/test_graph.lua"
local script_dir = script:match("^(.*)[/\\][^/\\]+$") or "."
local package_root = script_dir:gsub("[/\\]tests$", "")
if package_root == "tests" then
    package_root = "."
end

package.path = package_root .. "/src/?.lua;" .. package_root .. "/src/?/init.lua;" .. package.path

local graph = require("coding_adventures.graph")
local Graph = graph.Graph
local GraphRepr = graph.GraphRepr

local function assert_true(value, message)
    if not value then
        error(message or "expected true")
    end
end

local function assert_equals(expected, actual, message)
    if expected ~= actual then
        error((message or "values differ") .. string.format(" (expected=%s actual=%s)", tostring(expected), tostring(actual)))
    end
end

local function assert_list_equals(expected, actual, message)
    assert_equals(#expected, #actual, (message or "list length mismatch"))
    for i = 1, #expected do
        assert_equals(expected[i], actual[i], message or ("list mismatch at index " .. i))
    end
end

local function make_graph(repr)
    local g = Graph.new({ repr = repr })
    g:add_edge("London", "Paris", 300)
    g:add_edge("London", "Amsterdam", 520)
    g:add_edge("Paris", "Berlin", 878)
    g:add_edge("Amsterdam", "Berlin", 655)
    g:add_edge("Amsterdam", "Brussels", 180)
    return g
end

local function run_for_repr(repr)
    local g = make_graph(repr)
    assert_true(g:has_node("London"))
    assert_true(g:has_edge("London", "Paris"))
    assert_equals(5, g:size())
    assert_equals(520, assert(g:edge_weight("London", "Amsterdam")))

    local neighbors = assert(g:neighbors("Amsterdam"))
    assert_list_equals({ "Berlin", "Brussels", "London" }, neighbors)
    assert_equals(3, assert(g:degree("Amsterdam")))

    local bfs_order = assert(graph.bfs(g, "London"))
    assert_list_equals({ "London", "Amsterdam", "Paris", "Berlin", "Brussels" }, bfs_order)

    local dfs_order = assert(graph.dfs(g, "London"))
    assert_list_equals({ "London", "Amsterdam", "Berlin", "Paris", "Brussels" }, dfs_order)

    assert_true(graph.is_connected(g))
    assert_true(graph.has_cycle(g))

    local path = assert(graph.shortest_path(g, "London", "Berlin"))
    assert_list_equals({ "London", "Amsterdam", "Berlin" }, path)

    local mst = assert(graph.minimum_spanning_tree(g))
    assert_equals(4, #mst)
    assert_equals("Amsterdam", mst[1][1])
    assert_equals("Brussels", mst[1][2])
    assert_equals(180, mst[1][3])
end

run_for_repr(GraphRepr.ADJACENCY_LIST)
run_for_repr(GraphRepr.ADJACENCY_MATRIX)

do
    local g = Graph.new()
    g:add_edge("A", "B")
    g:add_edge("C", "D")
    local components = graph.connected_components(g)
    assert_equals(2, #components)
    assert_list_equals({ "A", "B" }, components[1])
    assert_list_equals({ "C", "D" }, components[2])
    assert_true(not graph.is_connected(g), "expected disconnected graph")
end

do
    local g = Graph.new()
    g:add_edge("A", "B")
    assert_true(g:remove_edge("A", "B"))
    assert_true(not g:has_edge("A", "B"))
    assert_true(g:remove_node("A"))
    assert_true(not g:has_node("A"))
end

do
    local g = Graph.new()
    assert_true(g:set_graph_property("name", "city-map"))
    assert_true(g:set_graph_property("version", 1))
    assert_equals("city-map", g:graph_properties().name)
    assert_equals(1, g:graph_properties().version)
    assert_true(g:remove_graph_property("version"))
    assert_equals(nil, g:graph_properties().version)

    g:add_node("A", { kind = "input" })
    g:add_node("A", { trainable = false })
    assert_true(g:set_node_property("A", "slot", 0))
    assert_equals("input", g:node_properties("A").kind)
    assert_equals(false, g:node_properties("A").trainable)
    assert_equals(0, g:node_properties("A").slot)
    assert_true(g:remove_node_property("A", "slot"))
    assert_equals(nil, g:node_properties("A").slot)

    g:add_edge("A", "B", 2.5, { role = "distance" })
    assert_equals("distance", g:edge_properties("B", "A").role)
    assert_equals(2.5, g:edge_properties("B", "A").weight)
    assert_true(g:set_edge_property("B", "A", "weight", 7.0))
    assert_equals(7.0, assert(g:edge_weight("A", "B")))
    assert_true(g:set_edge_property("A", "B", "trainable", true))
    assert_true(g:remove_edge_property("A", "B", "role"))
    assert_equals(true, g:edge_properties("A", "B").trainable)
    assert_equals(nil, g:edge_properties("A", "B").role)

    assert_true(g:remove_edge("A", "B"))
    local properties, err = g:edge_properties("A", "B")
    assert_equals(nil, properties)
    assert_equals("edge_not_found", err.type)
end

print("lua/graph tests passed")
